### PR TITLE
refactor(ios): optimize font registration

### PR DESCRIPTION
**Pull request by jahorton**
This PR addresses the concerns from https://github.com/keymanapp/keyman/issues/1492#issuecomment-2290743167 - the fact that our font-registration checking operations are quite unoptimized.  Querying the system for font families - and specifically, for the list of fonts of each font family - appears to be an expensive operation, so we should query this data as minimally as possible.

By restructuring the font-registration code to 'batch'-list all fonts that need to be checked, then doing a single pass through the system font data - rather than once per font - the operation should become markedly less costly.

As the original methods within `FontManager` are publicly visible from KeymanEngine - and thus are technically API that a framework consumer could already be calling - their purposes and signatures are preserved.  They'll use new methods where reasonable in order to improve their performance, and these new methods will all be marked private for now - no need to rush exposing them to consumers.

## User Testing

TEST_KEYBOARDS_WITH_SPECIAL_FONTS:  Using Keyman for iOS, install and test the following keyboards:
- Amharic - `gff_amharic`
- Khmer Angkor - `khmer_angkor`
    - Make special note of any font-related issues you observe...
    - both on the keyboard and in typed text
    - both in the app and when using the system keyboard

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/Classes/Resource Data/FontManager.swift [hunk 1]
```diff
@@ -26,15 +26,63 @@ public class FontManager {
...
+   * Checks for, and filters out, any fonts already registered on the system.
+   */
+  private func registerListedFonts(_ initialFontSet: Set<URL>) {
+    // If we are unable to read the font file's properties sufficiently,
+    // skip it.  We also don't need to register anything already registered.
+    var fontSet = initialFontSet.filter { !(cachedFont(at: $0)?.isRegistered ?? true) }
+    
+    // The prior line filters out any entries where cachedFont(at: $0) would be nil.
+    // Batch-lookups all fonts lacking cache-confirmation of prior registration.
+    var fontNamesToRegister = missingFonts(from: Set(fontSet.map { cachedFont(at: $0)!.name }))
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply ios/engine/KMEI/KeymanEngine/Classes/Resource Data/FontManager.swift:67]: This comment belongs in the code.

Heuristic: it should be a PR comment if it explaining something about the change between old and new code. If it is explaining something about how the new code works (without ref to the old code), then it really belongs in the code.

File: ios/engine/KMEI/KeymanEngine/Classes/Resource Data/FontManager.swift [hunk 2]
```diff
@@ -133,32 +168,54 @@ public class FontManager {
...
 
-  private func fontExists(_ fontName: String) -> Bool {
-    return UIFont.familyNames.contains { familyName in
-      UIFont.fontNames(forFamilyName: familyName).contains(fontName)
+  /**
+   * Queries the system for existing registrations for the specified fonts with a single batch run.
+   * Only fonts that could not be found will be returned within the result set.
+   */
+  private func missingFonts(from fontNames: Set<String>) -> Set<String> {
+    // Arrays are treated 'by value'; it's already a shallow copy.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Resource Data/FontManager.swift:198]: This comment doesn't belong -- it's about how Swift works, not about how our code works

```suggestion
```

**Review by mcdurdin — APPROVED** [review]

**Comment by dinakaranr** [comment]
# Test Results

* **TEST_KEYBOARDS_WITH_SPECIAL_FONTS (failed):**

I tested this issue with the attached "Keyman 18.0.90-alpha-test-12210" build on the iPhone 13 and iOS 17.4. Here is my observation.
1. Installed the "Keyman-18.0.90" file from testflight app and gave all permissions to the application.
2. Checked the "Enable Keyman as system-wide keyboard" and set the keyboard as the default keyboard box on the settings page.
3. Installed the "khmer_angkor" and "gff_amharic" keyboards.
4. Selected the "gff_amharic" as system keyboard.(1.5 version)
5. Open the Keyman app. Enable the "Predictions" to install the "Dictionary."
6. Added a sentence to the keyman notepad and chrome(search).
7. Enter a word and observe that the word is highlighted on the banner.
8. Press the "Spacebar" and then the word added.
9. The word prediction works on the banner and selection works correctly. "gff_amharic" works well.

I followed above steps for "khmer_angkor" keyboard. Observed below behavior
Added a word. select a word which highlighted on the banner.
Press the backspace key.
Here, [object, object] appears on the banner.
Please refer to the screenshot

**Comment by jahorton** [comment]
That error was a known issue (#12199) that has since been resolved (#12201).  I've gone ahead and pulled in the latest master, which includes the fix; could you try again and verify that the issue is resolved?

@keymanapp-test-bot retest all

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.98-alpha](https://keyman.com/downloads/releases/alpha/18.0.98)

**Pull request merged**